### ISSUE TITLE
Phel performance

### ISSUE THREAD

**Issue by mabasic** [phel-lang/phel-lang#247]
While building a website with Phel I noticed that the TTFB is very long for a simple hello phel page. 

It varies on my machine from 250ms to 300ms. 

Once I have added some logic to the website, TTFB increased and is now around 600ms.

I have tested running a simple index.php file `<?php echo "Hello, World!"; ?>` on the same machine and I get around 5ms TTFB.

Can we leave this issue open until the compile command arrives?

**Comment by jenshaase** [comment]
Can you add example code for "a simple phel page" and a "page with logic"?

However, the goal is to have two options:
1. A cache of compiled files for development. Currently with every request all code is compile from scratch. Not only your code but also Phel's core library and all other libraries you may be using. It would be better to only compile the changed code and load the unchanged code from a cache.
2. A compile command that compiles all your code to a distribution folder or a PHAR file. This can than be used in production.

**Comment by mabasic** [comment]
My server was losing disk space and after some analysis I found out that my `/tmp` directory was getting filled with files called like this: `__phelV5EVe6`. After opening one file I found this inside: 
 
```
<?php
// /home/mabasic/apps/mariobasic-n7/vendor/phel-lang/phel-lang/src/php/Runtime/../../phel/core.phel
// ;;AA8MA,CAAC,wCAAD,CAAA,CAAM,CAAC,uCAAD,CAAA,uEAAA,iCAAA,CAAN,CAAA
($GLOBALS["__phel"]["phel\\core"]["load"])(($GLOBALS["__phel"]["phel\\core"]["str"])("/home/mabasic/apps/mariobasic-/vendor/phel-lang/phel-lang/src/phel", "/core/basic-constructors.phel"));
```

Phel currently writes files to `/tmp` every time it compiles and then leaves the files behind.

Part of the output from `sudo ls /tmp`:

```
__phel7kD9MI  __phelFWpISU  __phelNGteJW  __phelV5DQBD  __phelcpOhI6  __phelkX5eqG  __phelsIASPq
__phel7kDJIx  __phelFWpIhm  __phelNGtgI8  __phelV5E0X9  __phelcpOo8c  __phelkX5z4k  __phelsIAYT8
__phel7kDg3Q  __phelFWpMeZ  __phelNGthKp  __phelV5E3nV  __phelcpP282  __phelkX6Y5Y  __phelsIAaqg
__phel7kDkmo  __phelFWpUFb  __phelNGtkqE  __phelV5EDkB  __phelcpP4Lf  __phelkX6n8z  __phelsIAfCo
__phel7kDlvR  __phelFWpdFe  __phelNGtnkL  __phelV5EHVk  __phelcpPJCf  __phelkX6ssK  __phelsIAjSo
__phel7kDoiH  __phelFWpigT  __phelNGtwcC  __phelV5EJ5d  __phelcpPtHY  __phelkX6tut  __phelsIAsxZ
__phel7kEEEP  __phelFWqJYK  __phelNGu5Va  __phelV5EVe6  __phelcpPueP  __phelkX6wp3  __phelsIAwag
```

It was taking over 6 GB on my server.

I could not remove the files using `sudo rm __phel*` because there were too many files for bash. I had to use:

```
cd /tmp
sudo su
find . -maxdepth 1 -name "__phel*" -print0 | xargs -0 rm
```

@jenshaase I'll send examples for "a simple phel page" and a "page with logic" later, but as I understand it now it all comes from Phel having to write these files to the disk and then reading from them.

**Comment by mabasic** [comment]
@jenshaase  I have seen in other languages that compile to Javascript like Rescript and Fable/F# that they have a "compile" command (compiler) which generates (in this case) .php files next to the .phel files. We could then see the php output from each phel file. I think that this would solve everything that you wrote in your response?

**Comment by mabasic** [comment]
Just an update. It takes 8 days to fill my disk with tmp files. From 45% usage to 90% usage on a 25gb drive in 8 days.

**Comment by HallofFamer** [comment]
How is Phel's performance relative to PHP? I think its a transpiler rather than an interpreter so it should not be much slower, but I am still curious about the speed. I am thinking about writing a Pharo Smalltalk transpiler into PHP similar to this one, and if the performance of Phel is close to raw PHP it will give me more hopes of that.

**Comment by jenshaase** [comment]
@HallofFamer Phel is compiled to PHP code. Currently this is done on the fly every time you get a request. However, we are working on `compile` command that compiles the Phel code directly to PHP. You can then execute the PHP files directly. Therefore, there are no big performance issues (but I can give you any number or test suites for that)

**Comment by HallofFamer** [comment]
@jenshaase 

Thats interesting, so basically it uses a transpiler that generates PHP code which incurs no overhead? Can you show an example of Phel code and the generated/transpiled equivalent PHP code so I can compare? Thanks.

**Comment by jenshaase** [comment]
Have a look at this folder: https://github.com/phel-lang/phel-lang/tree/master/tests%2Fphp%2FIntegration%2FFixtures

This is the integration test suite for the compiler. Each file contains Phel code and the corresponding PHP Code.

**Comment by jenshaase** [comment]
The compile command has been released in version 0.5.0
